FrontBundle/EventSubscriber/KernelExceptionSubscriber.php

```diff
@@ -50,57 +53,77 @@ public function __construct(
     {
         if ($event->getException() instanceof HttpExceptionInterface && '404' == $event->getException()->getStatusCode()) {
-            $currentSite = $this->siteRepository->findByAliasDomain($this->request->getHost());
 
-            if ($currentSite) {
-                $language = $this->getLanguage($currentSite);
-                $siteId = $currentSite->getSiteId();
-                $nodeId = ReadNodeInterface::ERROR_404_NODE_ID;
-                $node = $this->nodeRepository->findOnePublishedByNodeIdAndLanguageAndSiteIdInLastVersion($nodeId, $language, $siteId);
+            $this->setCurrentSiteInfo(
+                trim($this->request->getHost(), '/'),
+                trim($this->request->getPathInfo(), '/')
+            );
 
-                if ($node) {
-                    $html = $this->templating->render(
-                        'OpenOrchestraFrontBundle:Node:show.html.twig',
-                        array(
-                            'node' => $node,
-                            'parameters' => array('siteId' => $node->getSiteId(), '_locale' => $node->getLanguage())
-                        )
-                    );
-                    $event->setResponse(new Response($html, 404));
-                }
+            if ($this->currentSite && $this->currentLanguage && $html = $this->getCustom404Html()) {
+                $event->setResponse(new Response($html, 404));
             }
         }
     }
 
     /**
-     * Get the language to display the 404 page for $site
-     * If a language prefix maps the url, get the corresponding language
-     * Else get the language of the main alias
-     * 
-     * @param ReadSiteInterface $site
+     * Try to find and set the current site and language
      * 
-     * @return string
+     * @param string $host
+     * @param string $path
      */
-    protected function getLanguage(ReadSiteInterface $site)
+    private function setCurrentSiteInfo($host, $path)
     {
-        $targetAlias = null;
-        $path = trim($this->request->getPathInfo(), '/');
-        $host = $this->request->getHost();
+        $this->currentSite = null;
+        $this->currentLanguage = null;
+        $possibleSite = null;
+        $possibleAlias = null;
 
-        if ('' != $path) {
+        $matchingSites = $this->siteRepository->findByAliasDomain($host);
+
+        /** @var ReadSiteInterface $site */
+        foreach ($matchingSites as $site) {
             foreach ($site->getAliases() as $alias) {
-                if ($host == $alias->getDomain() && strpos($path . '/', $alias->getPrefix() . '/') === 0) {
-                    $targetAlias = $alias;
-                    break;
+                if ($alias->getPrefix() == '') {
+                    $possibleSite = $site;
+                    $possibleAlias = $alias;
+                } else {
+                    if ($host == $alias->getDomain() && strpos($path . '/', trim($alias->getPrefix(), '/') . '/') === 0) {
```

Comment:
I think that your check is not deep enough.

We wanted to find the longest prefix available.

For instance with the url  /en/domain/edit : 
 - site 1 : prefix en
 - site 2 : prefix en/domain

We should return the site 2